mir_eval/util.py

```diff
@@ -621,9 +621,120 @@ def match_events(ref, est, window):
+        Array of reference notes time intervals (onset and offset times)
+    ref_pitches: list, len=n
+        List of reference pitch values in Hertz
+    est_intervals : np.ndarray, shape=(m,2)
+        Array of estimated notes time intervals (onset and offset times)
+    est_pitches : list, len=m
+        List of estimated pitch values in Hertz
+    onset_tolerance : float > 0
+        The tolerance for an estimated note's onset deviating from the
+        reference note's onset, in seconds. Default is 0.05 (50 ms).
+    pitch_tolerance: float > 0
+        The tolerance for an estimated note's pitch deviating from the
+        reference note's pitch, in cents. Default is 50.0 (50 cents).
+    offset_ratio: float > 0
+        The ratio of the reference note's duration used to define the
+        offset_tolerance. Default is 0.2 (20%), meaning the offset_tolerance
+        will equal the ref_duration * 0.2 * 0.5 (0.5 since the window is
+        centered on the reference offset), or 0.05 (50 ms), whichever is
+        greater. Note: this parameter only influences the results if
+        with_offset=True.
+    with_offset: bool
+        If True, note offsets are taken into consideration in the matching,
+        where the offset tolerance depends on the offset_ratio parameter. If
+        False, offsets are ignored in the matching.
+
+    Returns
+    -------
+    matching : list of tuples
+        A list of matched reference and estimated notes.
+        ``matching[i] == (i, j)`` where reference note i matches estimate note
+        j.
+
+    """
+
+    # Compute the indices of feasible pairings
+    # hits = np.where(np.abs(np.subtract.outer(ref, est)) <= window)
+
+    # check for onset matches
+    onset_distances = np.abs(np.subtract.outer(ref_intervals[:, 0],
+                                               est_intervals[:, 0]))
+    onset_hit_matrix = onset_distances < onset_tolerance
+
+    # check for pitch matches
+    pitch_distances = np.abs(1200*np.log2(np.divide.outer(ref_pitches,
+                                                          est_pitches)))
+    pitch_hit_matrix = pitch_distances < pitch_tolerance
+
+    # check for offset matches if offset_ratio is not None
+    if with_offset:
+        offset_distances = np.abs(np.subtract.outer(ref_intervals[:, 1],
+                                                    est_intervals[:, 1]))
+        ref_durations = intervals_to_durations(ref_intervals)
+        offset_tolerances = 0.5 * offset_ratio * ref_durations
+        offset_tolerance_matrix = np.tile(offset_tolerances,
+                                          (offset_distances.shape[1], 1)).T
+        offset_hit_matrix = offset_distances < offset_tolerance_matrix
+    else:
+        offset_hit_matrix = np.ones_like(onset_hit_matrix)
+
+    # check for overall matches
+    note_hit_matrix = onset_hit_matrix * pitch_hit_matrix * offset_hit_matrix
+    hits = np.where(note_hit_matrix)
+
+    # Construct the graph input
```

Comment:
@bmcfee any objection to us putting the graph construction logic in `_bipartite_match`?  Or equivalently, do you ever see us having a use for `_bipartite_match` where we'd want to pass a graph directly instead of a hits matrix?